[Feature] 이의제기/반론 이벤트 구현

rhxfkg opened the pull request:
## 🔗 관련 이슈
Closes #15

## ✅ 작업 내용

### 💡개선 사항

#### 1. 이의제기/반론 등록 서비스 로직 구현
- **handleAttack**: 공격(이의제기) 등록
  - Phase/Turn 검증 (TEAM_A_ATTACK + A_ATTACK 또는 TEAM_B_ATTACK + B_ATTACK)
  - 팀별 ActiveBattleState에 저장 (teamA.attacks 또는 teamB.attacks)
  - PENDING 상태로 생성
  
- **handleDefense**: 반론(수비) 등록
  - Phase/Turn 검증 (TEAM_A_ATTACK + B_DEFENSE 또는 TEAM_B_ATTACK + A_DEFENSE)
  - **attackId 필드 제거**: 채택된 공격은 이미 전체 공개되므로 별도 참조 불필요
  - 팀별 ActiveBattleState에 저장 (teamA.defenses 또는 teamB.defenses)

#### 2. Phase/Turn 검증 헬퍼 메서드
- **canUserSubmitAttack**: 공격 가능 여부 검증
  - 중립 팀(NONE) 제한
  - Phase와 Turn 동시 검증 (예: TEAM_A_ATTACK + A_ATTACK일 때만 A팀 공격 가능)
  
- **canUserSubmitDefense**: 반론 가능 여부 검증
  - 중립 팀(NONE) 제한
  - Phase와 Turn 동시 검증 (예: TEAM_A_ATTACK + B_DEFENSE일 때만 B팀 반론 가능)

#### 3. 동료 코드 통합 (Phase/Turn 구조 반영)
- `BATTLE_PHASE`, `BATTLE_TURN`, `BATTLE_PLAYTIME`을 객체 형태로 변경
- `ActiveBattleState`에 `round`, `phase`, `turn`, `startedAt`, `expiredAt` 필드 추가
- `playTime` 타입 변경에 따른 수정 (`.time` 속성 사용)
- Mock 데이터 Phase 업데이트 (WAITING_FOR_START → OPINION_SHARE)

#### 4. 타입 시스템 개선
- **BattleDefense**: `interface`에서 `type alias`로 변경
  - attackId 필드 제거로 BattleDiscussion과 완전히 동일
  - `export type BattleDefense = BattleDiscussion`
  - ESLint `no-empty-object-type` 에러 해결

#### 5. WebSocket Gateway 이벤트 핸들러 (이전 커밋)
- **Battle:Attack** → **Battle:NewAttack**: 팀 룸 브로드캐스트
- **Battle:Defense** → **Battle:NewDefense**: 팀 룸 브로드캐스트
- **Battle:AttackVote**, **Battle:DefenseVote**: 투표 이벤트 (서비스 로직 미구현)

#### 6. Request DTO 설계 (이전 커밋)
- `AttackRequestDto`, `DefenseRequestDto` (attackId 제거), `AttackVoteRequestDto`, `DefenseVoteRequestDto`
- 기존 `BattleJoinRequestDto` 패턴 준수

<br />

## 📸 참고 사항

### 주요 구현 파일
- [battles.service.ts:229-315](backend/src/battles/service/battles.service.ts#L229-L315): 등록 로직 및 검증 헬퍼
- [battles.service.ts:163-197](backend/src/battles/service/battles.service.ts#L163-L197): ActiveBattleState 초기화 (필드 추가)
- [battles.const.ts](backend/src/battles/const/battles.const.ts): Phase/Turn 객체 구조 (동료 코드 반영)
- [battles.types.ts:62](backend/src/battles/types/battles.types.ts#L62): BattleDefense 타입 alias
- [battles.types.ts:75-90](backend/src/battles/types/battles.types.ts#L75-L90): ActiveBattleState 타입 확장

### Phase/Turn 검증 로직
공격과 반론은 **Phase + Turn 동시 검증**을 통해 정확한 시점에만 가능:
```typescript
// A팀 공격 가능: Phase가 TEAM_A_ATTACK이고 Turn이 A_ATTACK일 때
if (phase === 'TEAM_A_ATTACK' && turn?.status === 'A_ATTACK' && userTeam === BATTLE_TEAM.A)

// B팀 반론 가능: Phase가 TEAM_A_ATTACK이고 Turn이 B_DEFENSE일 때
if (phase === 'TEAM_A_ATTACK' && turn?.status === 'B_DEFENSE' && userTeam === BATTLE_TEAM.B)
```

### 테스트 결과
✅ 전체 테스트 통과: 69개 (기존 41개 + 신규 28개)
✅ 빌드 성공
DTO 검증, Gateway 이벤트, 서비스 로직 모두 테스트 커버
<br />

## 💬 질문사항 (고민했던 부분)
### 1. DTO 구조 선택
처음엔 payload 패턴({ payload: { authorId, content }, battleId })을 고려했으나, 기존 BattleJoinRequestDto 패턴과의 일관성을 위해 모든 필드를 하나의 DTO에 포함하는 방식으로 결정
### 2. async 키워드 제거
WebSocket 이벤트 핸들러와 서비스 메서드 모두 async 불필요 (현재는 인메모리 Map 사용)
ESLint @typescript-eslint/require-await 에러 방지
추후 Redis/DB 연동 시 async 추가 예정
### 3. attackId 필드 제거 결정
초기 구현에서는 반론이 어떤 공격에 대한 응답인지 명시하기 위해 attackId 필드를 포함했으나, 다음 이유로 제거:
채택된 공격은 이미 all.attacks로 전체 공개됨
반론은 채택된 공격에 대한 응답이므로 별도 참조 불필요
BattleDefense와 BattleDiscussion의 구조가 동일해져 타입 alias로 단순화 가능

rhxfkg committed:
feat: 이의제기/반론 WebSocket DTO 및 타입 정의 #15

rhxfkg committed:
feat: 이의제기/반론 WebSocket Gateway 구현 #15

WebSocket 기반 실시간 이의제기/반론 이벤트 핸들러 구현

주요 변경사항:
- Request DTO 생성 (AttackRequestDto, DefenseRequestDto, AttackVoteRequestDto, DefenseVoteRequestDto)
- DTO 검증 테스트 추가 (20개 테스트 통과)
- WebSocket Gateway 이벤트 핸들러 구현 (4개 이벤트)
  * Battle:Attack - 이의제기 등록
  * Battle:Defense - 반론 등록
  * Battle:AttackVote - 이의제기 투표
  * Battle:DefenseVote - 반론 투표
- Gateway 단위 테스트 추가 (8개 테스트 통과)
- socket.io 패키지 추가 (^4.8.1)

rhxfkg committed:
chore: 서비스 스텁 메서드 추가 - 빌드 에러 수정 #15

이의제기/반론 Gateway 구현을 위한 임시 서비스 메서드 스텁 추가
- handleAttack, handleDefense, handleAttackVote, handleDefenseVote
- 실제 구현은 다른 팀원이 진행 예정
- 현재는 'Not implemented yet' 에러 반환

HoonDongKang commented on backend/src/battles/const/battles.const.ts:
```diff
@@ -2,7 +2,9 @@ export const BATTLE_PHASE = {
   TEAM_A_ATTACK: 'TEAM_A_ATTACK',
+  TEAM_A_DEFENSE: 'TEAM_A_DEFENSE',
   TEAM_B_ATTACK: 'TEAM_B_ATTACK',
+  TEAM_B_DEFENSE: 'TEAM_B_DEFENSE',
```
`phase`와 `turn`에 대한 정의가 변경되었으니 참고해주세요!
제 코드에서 `BATTLE_PHASE`, `BATTLE_TURN` 참고 부탁드립니다.

rhxfkg committed:
refactor: Phase/Turn 구조 반영 #15

Danny의 feature/#53-battle-turn 브랜치 구조 반영
- BATTLE_PHASE: 객체 형태로 변경 (name, time 속성 포함)
- BATTLE_TURN: 추가 (A_ATTACK, B_ATTACK, A_DEFENSE, B_DEFENSE)
- BATTLE_PLAYTIME: 객체 형태로 변경 (name, time, rounds 속성 포함)
- BattlePhaseName, BattleTurn 타입 추가
- ActiveBattleState: phase, turn, round 필드 추가

rhxfkg committed:
feat: 이의제기/반론 등록 로직 구현 #15

- handleAttack: 공격 등록 서비스 메서드 구현
- handleDefense: 반론 등록 서비스 메서드 구현
- canUserSubmitAttack/Defense: Phase/Turn 검증 헬퍼 메서드 추가
- ActiveBattleState에 round, phase, turn, startedAt, expiredAt 필드 추가
- playTime 타입 변경에 따른 수정 (.time 속성 사용)
- WAITING_FOR_START → OPINION_SHARE로 변경

rhxfkg committed:
refactor: BattleDefense에서 attackId 필드 완전 제거 #15

- attackId 필드 제거 및 BattleDefense를 type alias로 변경
- DefenseRequestDto에서 attackId 제거
- Gateway 및 Service 메서드 시그니처 수정
- 테스트 수정 (attackId 관련 케이스 제거)

rhxfkg committed:
Merge branch 'dev' into feature/#15-discussion-event

Signed-off-by: rhxfkg <51471056+rhxfkg@users.noreply.github.com>

rhxfkg committed:
fix: dev 브랜치 머지 충돌 해결 #15

- battles.gateway.ts: handleDefenseVote 메서드 닫는 중괄호 추가
- battles.service.ts: playTime 타입 처리 수정 및 stub 메서드 반환 타입 명시
- handleAttackVote, handleDefenseVote 반환 타입을 BattleDiscussion, BattleDefense로 변경

U-S-jun reviewed (approved):
수고하셨습니다
